compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -916,6 +916,60 @@ impl<'body, 'tcx> VnState<'body, 'tcx> {
             .collect();
         let mut fields = fields?;
 
+        if let AggregateTy::Def(_, _) = &mut ty
+            && !fields.is_empty()
+        {
+            let copy_from = if let Value::Projection(first_pointer, ProjectionElem::Field(_, _)) =
+                *self.get(fields[0])
+            {
+                let mut copy_from_value = first_pointer;
+                while let Value::Projection(pointer, ProjectionElem::Deref) =
+                    *self.get(copy_from_value)
+                {
+                    copy_from_value = pointer;
+                }
+                if let Some(local) = self.try_as_local(copy_from_value, location) {
+                    let from_ty = self.local_decls[local].ty;
+                    if rvalue_ty == from_ty {
+                        Some((first_pointer, local, false))
+                    } else if Some(rvalue_ty) == from_ty.builtin_deref(false) {
+                        Some((first_pointer, local, true))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
```

Comment:
I had trouble understanding the logic here. Could you move it where we actually create the copied-from place?
I'd have in mind something like:
```rust
let mut projections = vec![];
let local = loop {
    if let Some(local) = self.try_as_local(copy_from_value, location) { break Some(local) }
    else if let Value::Projection(pointer, proj) = self.get(copy_from_value) { projection.push(proj); copy_from_value = pointer; }
    else { break None }
};
if let Some(local) = local {
    let place = Place { local, projection: tcx.mk_place_elems(projection) };
    *rvalue = Rvalue::Use(Operand::Copy(place));
}
```

Reply:
I mainly want to check if the copied type is the same first. I can't rely on `propagate_ssa`'s check because I changed the rvalue.